File: Command_Pattern/cpp/Command_Pattern.cc
#include <iostream>
#include <vector>
using namespace std;

class Command {
public:
    virtual void Execute() = 0;
};

class Light : public Command {
public:
    void Execute() {
        TurnOn();
    }
    void TurnOn() {
        cout << "Light turn on" << endl;
    }
};

class TV : public Command {
public:
    void Execute() {
        TurnOn();
    }
    void TurnOn() {
        cout << "tv turn on" << endl;
    }
};

class Switcher {
private:
    vector<Command*> CommandList;
public:
    void ExecandStore(Command* c) {
        c->Execute();
        CommandList.push_back(c);
    }
};

void test_command() {
    Switcher s;
    TV t;
    Light l;
    s.ExecandStore(dynamic_cast<Command*>(&t));
    s.ExecandStore(dynamic_cast<Command*>(&l));
}
int main() {
    test_command();
    return 0;
}

File: Command_Pattern/go/Command_Pattern.go
package Command_Pattern

import "fmt"

type Command interface {
	Execute()
}

type Light struct{}

func (this *Light) TrunOn() {
	fmt.Println("Light turn on")
}

func (this *Light) Execute() {
	this.TrunOn()
}

type TV struct{}

func (this *TV) TrunOn() {
	fmt.Println("tv turn on")
}

func (this *TV) Execute() {
	this.TrunOn()
}

type Switcher struct {
	CommandList []Command
}

func (this *Switcher) ExecandStore(c Command) {
	this.CommandList = append(this.CommandList, c)
	c.Execute()
}
